x/concentrated-liquidity/swaps_tick_cross_test.go

```diff
@@ -755,3 +877,96 @@ func (s *SwapTickCrossTestSuite) TestSwapOutGivenIn_Tick_Initialization_And_Cros
+
+	var (
+		tokenZeroDenom       = DefaultCoin0.Denom
+		tokeOneDenom         = DefaultCoin1.Denom
+		smallTokenOneCoinIn  = sdk.NewCoin(tokeOneDenom, sdk.NewInt(100))
+		smallTokenZeroCoinIn = sdk.NewCoin(tokenZeroDenom, sdk.NewInt(100))
+	)
+
+	// tests tick crossing behavior around min tick
+	// It first swaps to the min tick, then tries to swap below it and fails.
+	// At the same time, it validates that the liquidity and current tick are updated correctly.
+	// Additionally, it validates that it is still possible to swap right.
+	s.Run("to min tick", func() {
+		s.SetupTest()
+
+		poolId, _ := s.setupPoolAndPositions(tickSpacingOne, defaultTickSpacingsAway)
+
+		// Fetch pool
+		pool, err := s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, poolId)
+		s.Require().NoError(err)
+
+		// Compute tokenIn amount necessary to reach the min tick.
+		const isZeroForOne = true
+		tokenIn, _, _ := s.computeSwapAmounts(poolId, sdk.Dec{}, types.MinInitializedTick, isZeroForOne, shouldStayWithinTheSameBucket)
+
+		// Swap the computed large amount.
+		s.swapZeroForOneLeft(poolId, sdk.NewCoin(tokenZeroDenom, tokenIn.Ceil().TruncateInt()))
+
+		// Assert that min tick is crossed and liquidity is zero.
+		s.assertPoolTickEquals(poolId, types.MinCurrentTick)
+		s.assertPoolLiquidityEquals(poolId, sdk.ZeroDec())
+
+		// Assert that full range positions are now inactive.
+		s.assertPositionOutOfRange(poolId, types.MinInitializedTick, types.MaxTick)
+
+		// Refetch pool
+		pool, err = s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, poolId)
+		s.Require().NoError(err)
+
+		// Validate cannot swap left again
+		_, err = s.App.ConcentratedLiquidityKeeper.SwapExactAmountIn(s.Ctx, s.TestAccs[0], pool, smallTokenZeroCoinIn, tokeOneDenom, sdk.ZeroInt(), sdk.ZeroDec())
+		s.Require().Error(err)
+		s.Require().ErrorContains(err, types.InvalidAmountCalculatedError{Amount: sdk.ZeroInt()}.Error())
+
+		// Validate the ability to swap right
+		s.swapOneForZeroRight(poolId, smallTokenOneCoinIn)
+	})
+
+	// tests tick crossing behavior around max tick
+	// It first swaps to the max tick, then tries to swap beyond it and fails.
+	// At the same time, it validates that the liquidity and current tick are updated correctly.
+	// Additionally, it validates that it is still possible to swap left.
+	s.Run("to max tick", func() {
+		s.SetupTest()
+
+		poolId, _ := s.setupPoolAndPositions(tickSpacingOne, defaultTickSpacingsAway)
+
+		// Fetch pool
+		pool, err := s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, poolId)
+		s.Require().NoError(err)
+
+		// Compute tokenIn amount necessary to reach the max tick.
+		const isZeroForOne = false
+		tokenIn, _, _ := s.computeSwapAmounts(poolId, sdk.Dec{}, types.MaxTick, isZeroForOne, shouldStayWithinTheSameBucket)
+
+		// Swap the computed large amount.
+		s.swapOneForZeroRight(poolId, sdk.NewCoin(tokeOneDenom, tokenIn.Ceil().TruncateInt()))
+
+		// Assert that max tick is crossed and liquidity is zero.
```

Comment:
can max tick be crossed?

Reply:
Same comment here, it seems like the verbiage is confusing of "cross" vs "entering". It feels like "crossing" any max should not be possible.

Reply:
It is correct. We do cross it.

Since we can only have upper ticks of positions be initialized on the max tick, the liquidity net amounts on them are always negative.

Therefore, when swapping one for zero and crossing a max tick to be "within in", we always end up with a liquidity of zero.

Ported this context here: [e196c14](https://github.com/osmosis-labs/osmosis/pull/5581/commits/e196c14d1584f8a5c4ebbcec3d7f147ccdac25bd)